nexus/src/app/sagas/snapshot_create.rs

```diff
@@ -1958,66 +1999,132 @@ mod test {
-        // since the snapshot is created from a disk.
-        //
-        // Unfortunately, for our idempotency checks, checking for a "clean
-        // slate" gets more expensive when we need to compare region allocations
-        // between the disk and the snapshot. If we can undo the snapshot
-        // provisioning AND delete the disk together, these checks are much
-        // simpler to write.
-        //
-        // So, in summary: We do some odd indexing here...
-        // ... because we re-create the whole DAG on each iteration...
-        // ... because we also delete the disk on each iteration, making the
-        // parameters invalid...
-        // ... because doing so provides a really easy-to-verify "clean slate"
-        // for us to test against.
-        let mut n: usize = 0;
-        while let Some(node) = dag.get_nodes().nth(n) {
-            n = n + 1;
-
-            // Create a new saga for this node.
-            info!(
-                log,
-                "Creating new saga which will fail at index {:?}", node.index();
-                "node_name" => node.name().as_ref(),
-                "label" => node.label(),
-            );
 
-            let runnable_saga =
-                nexus.create_runnable_saga(dag.clone()).await.unwrap();
-
-            // Inject an error instead of running the node.
-            //
-            // This should cause the saga to unwind.
-            nexus
-                .sec()
-                .saga_inject_error(runnable_saga.id(), node.index())
-                .await
-                .unwrap();
-            nexus
-                .run_saga(runnable_saga)
-                .await
-                .expect_err("Saga should have failed");
+        // As a concession to the test helper, make sure the disk is gone
+        // before the first attempt to run the saga recreates it.
+        delete_disk(client, PROJECT_NAME, DISK_NAME).await;
+
+        // The no-pantry variant of the test needs to see the disk attached to
+        // an instance. Set up an IP pool so that instances can be created
+        // against it.
+        if !use_the_pantry {
+            populate_ip_pool(
+                &client,
+                "default",
+                Some(
+                    IpRange::try_from((
+                        Ipv4Addr::new(10, 1, 0, 0),
+                        Ipv4Addr::new(10, 1, 255, 255),
+                    ))
+                    .unwrap(),
+                ),
+            )
+            .await;
+        }
 
-            delete_disk(client, PROJECT_NAME, DISK_NAME).await;
-            verify_clean_slate(cptestctx, &test).await;
-            disk_id =
-                create_disk(client, PROJECT_NAME, DISK_NAME).await.identity.id;
+        crate::app::sagas::test_helpers::action_failure_can_unwind::<
+            SagaSnapshotCreate,
+            _,
+            _,
+        >(
+            nexus,
+            params,
+            || {
```

Comment:
This function is documented as:

> /// - `generate_params`: A callback called at the beginning of each loop
> ///   iteration that returns a future that yields the saga parameters to use for
> ///   that loop iteration.

This seems to kinda be treading into the realm of `after_saga`, by actually modifying state.

IMO it's kinda surprising to have a "generate_params" function also act as a "test setup" function -- if we want this, should we update the name accordingly?

Reply:
Renamed in e5030eb09.